### PR TITLE
perf(hf-transformers): reuse EAGLE-3 draft mask buffer

### PR DESCRIPTION
**Pull request by beomsu-mobilint**
## 개요

- EAGLE-3 draft-model forward의 fp32 attention mask 재구성을 없애고, 지연 할당된 `(1, 1, MAX_TGT, MAX_KV)` 버퍼를 재사용하도록 변경. draft `forward`가 넘기던 `torch.ones` + `expand_mask` no-op 체인도 함께 제거.
- 실측 same-process decode TPS가 4개 config 모두 개선. 특히 W8 greedy `+10.09%` (11.06 → 12.18 tok/s). Greedy 200-token 출력은 W8/W4V8 두 리비전 모두 byte-identical.
- 변경 범위: `mblt_model_zoo/hf_transformers/utils/eagle3/eagle3_utils.py` + `MobilintEagle3GenerationMixin._prepare_eagle3_cache`에서 `reset_draft_mask_state()` 호출 1건.

### 1단계 — Pre-change 프로파일 (mask cluster 발견)

`mobilint/EAGLE3-Llama-3.1-8B-Instruct` W8 greedy 200-token decode에 cProfile 3회 반복. 결과:

- **NPU 대기 (`qbruntime.Model._infer`) = tottime의 94.34%** — 물리적 하한.
- Python 측 프레임은 모두 개별 <1% tottime.
- **주목**: draft path의 `_prepare_decoder_attention_mask` + `make_causal_mask` + `expand_mask` + `convert_attention_mask_to_numpy` 클러스터가 tottime 0.64%에 불과했으나, 매 draft-depth step에서 fp32 `(1, 1, tgt, kv)` 할당이 발생 (200-token decode 기준 ~348회). **cProfile tottime이 잡지 못한 malloc allocator pressure**가 있을 가능성이 큼.

Pre-change 시점의 잘못된 후보들도 이 단계에서 empirical하게 걸러냄:
- `qbruntime.model_variant_handle.get_model_input_shape` — cumtime 5.5%로 보였으나 이는 pybind11 boundary에서 cProfile이 유발한 자체 오버헤드였음. 실제 tottime은 0.44%, 실측 회수 가능 wall-time은 0.03%로 노이즈 아래. 별도 실험(Python-side memoize monkey-patch)에서 REJECT.
- `evaluate_posterior`의 fancy-indexed argmax — PR #107이 이미 flat argmax로 해결했음을 재확인.

### 2단계 — Phase 0 tottime gate

Task 스코프를 확정하기 전에, mask cluster의 tottime 실측치를 다시 계산:

| Run | Core-fn tt | Native-helper tt | draft-path tt 총합 | decode wall | tottime % |
|-----|-----------:|-----------------:|-------------------:|------------:|----------:|
| 1   | 0.0530 s   | 0.0522 s         | 0.1052 s           | 16.028 s    | **0.656%** |
| 2   | 0.0533 s   | 0.0526 s         | 0.1059 s           | 16.018 s    | **0.661%** |
| 3   | 0.0532 s   | 0.0522 s         | 0.1054 s           | 17.275 s    | **0.610%** |

평균 **0.642%**. `[0.5%, 1%)` 구간 → "**축소된 상한 명시하고 진행**" 결정. cProfile tottime만 보면 회수 상한이 노이즈 근처지만, allocator-pressure 각도가 명확했으므로 진행.

### 3단계 — Post-change 재프로파일 (확인 + 시리즈 종료)

이번 커밋 적용 후 같은 방법론으로 재프로파일한 결과:

- **NPU 대기 94.34% → 96.11%** (~1.8pp Python 측 CPU 작업 회수 확인).
- Mask 클러스터가 draft 경로에서 완전히 제거됨:

| Frame                | 이전 ncalls | 현재 ncalls | 상태 |
|----------------------|------------:|------------:|-----|
| `make_causal_mask`   | 396         | 60          | draft 제거 |
| `torch.ones`         | 407         | 120         | draft 제거 |
| `torch.rsub`         | 407         | 60          | draft 제거 |
| `expand_mask`        | 407         | 60          | draft 제거 |
| `torch.arange`       | 570         | 237         | draft 제거 |
| `torch.full`         | 512         | 178         | draft 제거 |
| `Tensor.to`          | 4684        | 3358        | −1326회 (fp32 mask 할당 흔적 제거) |
| `masked_fill_` (inplace) | 396     | 696         | 이번 변경의 in-place 채움 증가 |

- 나머지 사전 후보 (draft/base numpy 변환 체인, `last_hidden.repeat`, depth-loop `torch.cat`) 모두 tottime 0.02~0.16%로 임계 미달. Allocation-pressure 각도 없음.
- **이 저장소 내부에서 남은 유의미(≥0.5% wall-time) 최적화 여지 없음**을 데이터로 확인. 시리즈 종료.

## 성능

Same-process (`--repeat 3`) — MXQ cross-process non-determinism 감쇠 (`CLAUDE.md` 지침):

| Config      | before mean ± sd     | after mean ± sd      | Δ TPS   |
|-------------|---------------------:|---------------------:|--------:|
| W8 greedy   | 11.064 ± 0.015 tok/s | 12.180 ± 0.014 tok/s | +10.09% |
| W4V8 greedy | 13.490 ± 0.044 tok/s | 13.712 ± 0.030 tok/s |  +1.64% |
| W8 t=0.7    | 11.058 ± 0.178 tok/s | 11.744 ± 0.941 tok/s |  +6.20% |
| W4V8 t=0.7  | 13.775 ± 0.226 tok/s | 14.316 ± 1.581 tok/s |  +3.93% |

Per-token CPU (same-process):

| Config      | before ms/tok  | after ms/tok   | Δ                |
|-------------|---------------:|---------------:|-----------------:|
| W8 greedy   | 3.026 ± 0.086  | 2.795 ± 0.069  |  −0.232 (−7.6%)  |
| W4V8 greedy | 5.374 ± 0.218  | 4.579 ± 0.078  |  −0.795 (−14.8%) |
| W8 t=0.7    | 5.543 ± 0.318  | 4.963 ± 0.362  |  −0.580 (−10.5%) |
| W4V8 t=0.7  | 6.518 ± 0.037  | 5.854 ± 0.511  |  −0.664 (−10.2%) |

Cross-process (n=3 독립 프로세스): 4.45% MXQ non-determinism 노이즈 안에 모두 들어옴. same-process 신호와 정합하며 primary evidence가 아님.

**실측 W8-greedy TPS 이득이 Phase 0의 tottime 예측(~0.64%)을 크게 상회한 이유**: cProfile tottime은 함수별 self-time만 잡음. 하지만 매 generate cycle마다 발생하던 ~348회의 fp32 mask 할당은 malloc/free allocator에 압력을 만들었고, 이 압력은 특정 함수에 귀속되지 않은 채 동일 프로세스에서 반복되는 generate 호출 성능을 훼손하고 있었음. 이 커밋은 그 압력을 근원에서 제거함.

측정 커맨드:
```
mblt-model-zoo tps measure \
  --model mobilint/EAGLE3-Llama-3.1-8B-Instruct \
  --revision {W8|W4V8} --temperature {0|0.7} \
  --input-mode synthetic-text --decode 200 \
  --prompt-text "What is transformer model?" \
  --device cpu --print-output
```

## 정확성

- Greedy: `--temperature 0`으로 생성한 200-token 텍스트가 before/after byte-identical. W8 1055 B, W4V8 1061 B.
- `tokens_per_step` 및 `draft_accept_ratio` greedy 기준 완전 동일: 3.242 / 38.82% (W8), 3.045 / 35.76% (W4V8).
- Sampling `tokens_per_step` drift는 same-process n≥10 runs 기준 ±5% 안: W8 t=0.7 `+3.75%`, W4V8 t=0.7 `+3.56%`.
- 재사용 버퍼 경로를 pre-change `make_causal_mask` + `expand_mask` 경로와 대조 검증: prefill + 5개 tree-expansion depth × 여러 prompt length + `reset_draft_mask_state()` 후 다른 shape 재실행 케이스 전부에서 `torch.equal` 성립. `convert_attention_mask_to_numpy` 출력도 `np.array_equal`.

## 구현 노트

- `_draft_mask_buf`는 `(1, 1, tgt, kv)` fp32 버퍼로, 첫 `_prepare_decoder_attention_mask` 호출 시 지연 할당되고 필요시 확장됨. Past-KV 컬럼은 항상 0. 매 호출은 `(tgt, tgt)` tree 영역만 재작성하고, 이전 tree 영역은 `_draft_mask_prev_span` tracker로 zero 복원.
- `reset_draft_mask_state()`는 이전 span tracker를 초기화. `MobilintEagle3GenerationMixin._prepare_eagle3_cache`에서 호출하여 이전 `generate()` 호출의 stale write가 다음 run의 past-KV 컬럼을 오염시키지 않게 함.
- Base-model mask path (`eagle3_utils.py:504-528`)는 손대지 않음. Base는 iteration당 1회만 fire되며 hot cluster가 아님.
- non-`bool` `attention_mask` (현재 caller들은 사용하지 않음)는 원래의 `expand_mask` + add slow path로 폴백. Fast path가 예상하지 못한 shape을 조용히 처리하는 일 방지.

## 테스트 플랜

- [x] `pytest tests/transformers/text_generation/eagle3/` — 136 passed, 1 deselected (pre-existing, 무관)
- [x] W8 및 W4V8 greedy 200-token 출력 byte-identical
- [x] Same-process TPS 매트릭스: 2 revisions × 2 temperatures × before/after
- [x] Cross-process TPS 매트릭스 sanity check
- [x] Mask 등가성 프로브: prefill + 5 tree depths + reset-후-reshape
- [x] Post-change 재프로파일로 mask cluster 제거 및 NPU 대기 96.11% 확인

### CODE REVIEW

File: mblt_model_zoo/hf_transformers/utils/eagle3/eagle3_utils.py [hunk 1]
```diff
@@ -714,30 +743,98 @@ def _prepare_decoder_attention_mask(
...
+
+        combined_attention_mask = buf[:, :, :target_length, :kv_length]
+
+        # In the draft path, callers only ever pass an all-visible bool
+        # ``attention_mask`` (via the default in ``forward``) or ``None``;
+        # ``expand_mask`` on all-ones produces an all-zero contribution that
+        # is added to ``combined_attention_mask`` as a no-op. Fall back to
+        # the slow path only when a caller supplies a non-trivial mask
+        # (dtype-driven check, no device sync).
+        if attention_mask is not None and attention_mask.dtype != torch.bool:
```
[the remark this answers is not in the document]
Comment (beomsu-mobilint) [reply mblt_model_zoo/hf_transformers/utils/eagle3/eagle3_utils.py:822]: Addressed in a57994c: bool masks with False entries now raise ValueError explicitly, so silent wrong logits are impossible. All-True bool masks (the internal default) keep the fast path.

**Comment by beomsu-mobilint** [comment]
@codex

File: mblt_model_zoo/hf_transformers/utils/eagle3/eagle3_utils.py [hunk 2]
```diff
@@ -714,30 +743,104 @@ def _prepare_decoder_attention_mask(
...
+        # fast path would otherwise silently ignore the masked positions.
+        # Non-bool masks go through the slow ``expand_mask`` + add path.
+        if attention_mask is not None:
+            if attention_mask.dtype != torch.bool:
+                expanded_attn_mask = expand_mask(
+                    attention_mask, torch.float32, tgt_len=target_length
+                ).to(hidden_states.device)
+                combined_attention_mask = combined_attention_mask + expanded_attn_mask
+            elif not attention_mask.all():
+                raise ValueError(
```
[the remark this answers is not in the document]
Comment (beomsu-mobilint) [reply mblt_model_zoo/hf_transformers/utils/eagle3/eagle3_utils.py]: Addressed in 58e983b: the .all() validation now runs before the causal block write, so a partial bool mask raises without any buffer mutation. Verified with a snapshot-then-raise-then-torch.equal check that the buffer state is untouched on the error path.

**Comment by beomsu-mobilint** [comment]
@codex

File: mblt_model_zoo/hf_transformers/utils/eagle3/eagle3_utils.py [hunk 3]
```diff
@@ -714,30 +743,104 @@ def _prepare_decoder_attention_mask(
...
+            buf is None
+            or buf.size(0) < batch_size
+            or buf.size(-2) < target_length
+            or buf.size(-1) < kv_length
+            or buf.device != hidden_states.device
+        )
+        if need_alloc:
+            new_batch = max(batch_size, buf.size(0) if buf is not None else 1)
+            new_tgt = max(target_length, buf.size(-2) if buf is not None else 0, 16)
+            new_kv = max(kv_length, buf.size(-1) if buf is not None else 0, 128)
```
[the remark this answers is not in the document]
Comment (beomsu-mobilint) [reply mblt_model_zoo/hf_transformers/utils/eagle3/eagle3_utils.py]: Addressed in b3f4e8e: buffer growth is now geometric on both tgt and kv dimensions (`max(current * 2, needed, floor)`). Amortized reallocations drop from O(N) to O(log N) across a decode; correctness invariants (past-KV columns zero, prev-span cleanup covers all writes) are unchanged.

**Comment by beomsu-mobilint** [comment]
@codex
